### DISCUSSION TITLE
Support both language and diff highlighting

### DISCUSSION THREAD

**Discussion by ninofiliu** [github-linguist/linguist#5758] [Q&A] [answered]
Is there any way to support both language highlighting on top of diff highlighting?

This works (ie `goodbye` has a red background and `hello` has a green background)

    ```diff
    - goodbye
    + hello
    ```

```diff
- goodbye
+ hello
```

This works (ie `const` is in red, `a` is in white, and `10` is in blue)

    ```js
    const a = 10;
    ```

```js
const a = 10;
```

But this doesn't (ie `10` is not in blue with a red background). I tried with `diff&js`, `diffjs`, `diff-js`, `jsdiff`, `js-diff`, `js&diff`, etc.

    ```diff&js
    - const a = 10;
    + const a = 20;
    ```

```diff&js
- const a = 10;
+ const a = 20;
```

I know this can be done because we have both diff and language highlighting in pull requests already

![image](https://user-images.githubusercontent.com/29477588/152210064-407ee412-9f08-48d4-b085-1e699c4fa0d5.png)

but apparently there doesn't seem to be a way to do it in github-flavored markdown, am I right? Is it a feature that is being developed right now? If not, could I contribute to this repo to add this feature?

According to my research, it's not possible at the time, but perhaps I am missing something:

- [Stack Overflow - Diff syntax highlighting in Github Markdown](https://stackoverflow.com/q/40883421/8186898) -> not possible
- [Syntax Highlighting Combine diff and xxx](https://stackoverflow.com/q/41405248/8186898) -> not possible
- [Supported languages](https://github.com/github/linguist/blob/master/vendor/README.md) -> no explicit mention of diff + xxx

**Comment by lildude** [comment, marked as answer]
> Is there any way to support both language highlighting on top of diff highlighting?

No, and it's not really up to Linguist either. Inline rendering is handled by https://github.com/github/markup which in turn uses Linguist when necessary. 

I can't comment on if this feature is being developed as I'm not part of that side of things, though I know that project is stagnant right now. I think your question may also fall into the issue already open there (I've not read it all 😁) https://github.com/github/markup/issues/1440

**Reply by ninofiliu** [reply]
Thanks! The referred threads do not address diff+language in particular but I'll make do with the markup repo pointer you have me

**Reply by Alhadis** [reply]
I've been requesting this feature for years (including more recently), so don't hold your breath waiting on a change. I've actually seen this implemented on other markdown-powered sites (IIRC, it was either GitLab or BitBucket). The syntax looked like this:

~~~markdown
```diff js
 let fileIcon = document.querySelector("li.file-entry > span.icon");
+fileIcon.classList.remove("icon-file-text");
 const disposable = addIconToElement(fileIcon, "/path/to/file.txt");
```
~~~

Currently, the closest we have are `suggestion` blocks, which actually diff highlighted code [_perfectly_](https://github.com/github/linguist/pull/5811#discussion_r818461370):

<table width="700"><thead><tr><th
>Source</th><th
>Output</th>
</tr></thead><tbody><tr><td valign="middle" width="350">

~~~markdown

```suggestion
ace_mode: pascal
codemirror_mode: pascal
codemirror_mime_type: text/x-pascal
```

~~~

</td><td valign="top" width="540">

<img width="540" src="https://user-images.githubusercontent.com/2346707/171088142-fef8dc38-c9f4-4c34-b724-6a3b2a1d2ff4.png" alt="Figure 1: Rendered output" />

</td></tr></tbody></table>

So it's not out of the realm of possibility that GitHub already have this implemented somewhere… they just need to hook it up to whatever code's responsible for sprucing up fenced code-blocks. 😉

**Reply by mderazon** [reply]
The syntax you mentioned looks like the best option to me. However take into consideration that it can mess up formatters like Prettier, that currently know how to auto-format code inside the fenced code blocks in Markdown which is very handy. That inline diff syntax characters (`+` / `-`) can mess it up.
So I guess not so simple...

[Playground Link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAFATnGMCWd0AEAZhOgLYCGMAzgTABZzEQA2LEA7jlAOYGQATJgCN2YANa0OcNkgA6UAAbKAVtQUssxHJoCSkKAQC8BARDABXMghgA6AI4X8ATwDKMuGBikAFHJAsOLZEOnAAtDbozgQAfATUAA4UULY4Bv4AlADcCgR5+fkhega2YCwU1NQAMjjUdphkEABucH4gadBhReHwAB4wmTmGBQUGdaa1CRDUFKJMJhQCAvrQACoQAKKa1rA+3StQADQE-gD0SQyn3qfdtjD9gwrKigoKAOqMhooA1IoTRER4s5YBRegRagRFkIBMccDBiBQdLRvMxyFRIbQ4QByWhQCDwigEJoUQICAiKNR-ajAmCgmDyJSqdRQTTw-YGYymcxWGwOJxRdyaLy+fyBYKhCKwKKxeJJFIdKCPYYjbTFaClcqVGp1WwNZqtfwKroSvoDEDZBTfFX8aDjASTaazTScqEHdZbOA7GB7UIHY5nC70K4QG6hO4Pc1DZ4gQ4gCAJXC25CgCjodCcVCphDUZAgEkcCjOHOx4ToCgSLCuCjWGpQODIIgk6hwEtliswVxJMDcHjIGDoJyxz3COBLUdVZI8CwUHhwABipEo2B7uYoFm8MZA9BgZBYHzhcES5bg7ig1DhOCacOcubAlU33Gb6BgGBnlAbTZbIDUvVcPc0ACKFj4vWSCNiwzaxkk6BPrmlDoOIZgcFAm4JOg3AwG8OACAwyAABwAAxQemzZvGWCS5mhh74C0m6OCBGDxjmKAVGEdajqOm6YI4OCYK+PDvmBn6xs2ZA4H2A5fuevCaAAgtg6HCOucCoPgtageBkEgNJPCAcB8AfhBX60sIWE4fQyAAEyxv2iKBLwADCEBkIJICHgArJuFjNqsszMZpX5NE4uhQEIsCuGA6EJrJoWuDAziaIZzYAL7JUAA)

Still, i'd rather have this feature and Prettier fail to auto format in this case than not having this feature at all!

**Reply by hb20007** [reply]
I tried opening an Issue (https://github.com/github/markup/issues/2025) on the  https://github.com/github/markup repository and they said their repository is not responsible for it and that [github/linguist](https://github.com/github/linguist) handles syntax highlighting.

**Reply by hb20007** [reply]
@lildude What do you suggest for requesting this feature? Perhaps opening an Issue on this repo?

**Reply by lildude** [reply]
Linguist is only responsible for providing the grammars used by the syntax highlighting engines (there are two; both are internal repos). The ability to specify two options in inline code blocks and have them used by the highlighting engine would need to be implemented in markup, like the suggestion box does (Linguist isn't responsible for that functionality).
